0022-lnd-plan-of-attack.adoc

```diff
@@ -0,0 +1,448 @@
+
+We can fetch all payments (including failed ones) with the following GET request
+
+[source=http]
+----
+GET /v1/payments?include_incomplete=true HTTP/1.1
+Host: LND_HOST:LND_PORT
+----
+
+Documentation: https://api.lightning.community/rest/index.html#v1-payments
+
+. Advantages:
+- Simple HTTP GET request
+
+. Disadvantages:
+- We have to poll regularly
+
+====== Using gRPC
+
+I couldn't find any subscription-based RPC call for outgoing payments but there is the RPC equivalent of the above HTTP call: `listpayments`
+
+Documentation here: https://api.lightning.community/#listpayments
+
+. Advantages
+- No clear advantages
+
+. Disadvantages
+- Need to integrate gRPC and protobuf code generation in the code base
+
+==== How to authenticate with LND?
+
+LND provides macaroons that allow users to very precisely specify the permissions a specific client has.
+Our users will have to provide a macaroon to cnd that we can use to authenticate against it.
+Ideally, we can just tell the user to put the macaroon at a specific place (inside the cnd config folder) so we can automatically discover it.
+
+==== What are the actions we need to emit on cnd's REST API?
+
+Two make the above described protocol we happen, we need to extend cnd's REST API.
+In particular, we need to add three actions:
+
+. `lnd-send-payment`
+. `lnd-add-hold-invoice`
+. `lnd-settle-invoice`
+
+I choose to prefix those with `lnd` instead of `lightning` because "hold invoices" are a client feature of Lightning and don't exist on the protocol level.
+This set of actions allows us to keep cnd's relationship with lnd read-only.
+
+===== `lnd-send-payment`
+
+To send a payment through lnd, this action needs to contain the following parameters:
+
+. recipient lnd pubkey
+. amount
+. secret-hash
+. final-cltv-delta:
+The CLTV (Check-Lock-Time-Verify) delta of the HTLC in the last hop of the payment routing.
+This value is conceptually equivalent with the on-chain expiry value (it is just relative in this case).
+
+===== `lnd-add-hold-invoice`
+
+Adds a new hold invoice to lnd.
+This action could return up to four parameters:
+
+. amount
+. secret-hash
+. expiry:
+This allows us to specify, for how long we are willing to accept a payment for this invoice.
+The default here is 3600 seconds, e.g. 1 hour.
+We should return sensible values here, depending on the roles and the protocol that is used on the other chain.
+If this timeout is too short, Bob might not be able to pay the invoice in the X-HALight case even though we would like to proceed with the swap.
+For the HALight-X case, this would allow Bob to specify a time window up until he would like Alice to start the swap.
+In this case, the timeout should not be too long, otherwise the absolute expiry of his HTLC (assuming a HAN-protocol) is not enough for a secure swap.
+. memo
+We can return a description that says something like "COMIT atomic swap {swap-id} with {peer-id}" to make identifying invoices of swaps easier for the user.
```

Comment:
FYI, max size is 639 bytes (I guess we are fine) https://github.com/lightningnetwork/lightning-rfc/blob/master/11-payment-encoding.md#tagged-fields